File: Caribbean Online Judge COJ/1683 - DPA/1683 - DPA.cpp
#include <cstdio>
using namespace std;
/*
* Autor: Angel Jaime
* 2732 - Coco-Bits and the Coco-Strings. COJ
*
* Problema de simulacion. Solamente hay que leer el numero de casos a procesar, y para cada caso
* realizar una sumatoria desde 1 hasta N - 1, de todos aquellos numeros entre los que N sea divisible,
* si esta sumatoria es menor a N, entonces imprimimos "Deficient", si es igual, imprimos "Perfect", de
* lo contrario imprimimos "Abundant". Aca en el caso de C++ se implemento una optimizacion de factores primos
* en la sumatoria, pero el cometido es igual, se puede resolver de la misma manera que en la solucion de Java
*/
int c, num;

int sumOfDivisors(int num) //Metodo para la sumatoria de divisores
{
    int p, i, prod = 1;
    for(i = 2 ; i * i <= num ; i++)
    {
        p = 1;
        while(num % i == 0)
        {
            p = (p * i) + 1;
            num = num / i;
        }
        prod *= p;
    }
    if(num > 1)
    {
        prod *= 1 + num;
    }
    return prod;
}

int main()
{
    scanf("%d", &c);
    while(c--) //Para cada caso a evaluar
    {
        scanf("%d", &num); //Leemos valor de N y evaluamos mandando llamar el metodo de la sumatoria directamente
        printf(((sumOfDivisors(num) - num) < num) ? "Deficient\n" : (((sumOfDivisors(num) - num) == num) ? "Perfect\n" : "Abundant\n"));
    }
    return 0;
}

File: Caribbean Online Judge COJ/1683 - DPA/1683 - DPA.java
import java.util.Scanner;
/*
* Autor: Angel Jaime
* 1683 - DPA. COJ
*
* Problema de simulacion. Solamente hay que leer el numero de casos a procesar, y para cada caso
* realizar una sumatoria desde 1 hasta N - 1, de todos aquellos numeros entre los que N sea divisible,
* si esta sumatoria es menor a N, entonces imprimimos "Deficient", si es igual, imprimos "Perfect", de
* lo contrario imprimimos "Abundant"
*/
class Main
{
	public static void main (String[] args) throws java.lang.Exception
	{
		Scanner s = new Scanner(System.in);
		int N, i, sum, nCases = s.nextInt();
		
		while(nCases != 0)
		{
			N = s.nextInt(); //Leemos valor de N
			sum = 0; //Para cada caso reiniciamos el valor de la sumatoria
			for(i = 1 ; i < N ; i++) //Desde 1 hasta N - 1. OJO, no iniciar desde 0 o se causara indeterminacion matematica
			{
				if(N % i == 0) //Si N es divisible entre el iesimo valor actual (Si el modulo '%' o resto de la division es igual a 0)
				{
					sum += i; //Sumamos
				}
			}
			if(sum < N) System.out.println("Deficient"); //Aqui ya solamente se evalua si la suma es menor a N...
			else if(sum == N) System.out.println("Perfect"); //si es igual....
			else System.out.println("Abundant"); //Si es mayor
			nCases--; //Decrementamos numero de casos
		}
	}
}
